src/box/lua/experimental/connpool.lua

```diff
@@ -0,0 +1,59 @@
+local config = require('config')
+local checks = require('checks')
+local netbox = require('net.box')
+
+local connections = {}
+
+local function connect_using_config(instance_name, opts)
+    checks('string', {
+        wait_connected = '?boolean|number',
+        fetch_schema = '?boolean',
+    })
+    opts = opts or {}
+
+    local conn = connections[instance_name]
+    if conn == nil or conn.state == 'error' or conn.state == 'closed' or
+       conn.space == nil and opts.fetch_schema then
+        local uri_opts = {
+            instance = instance_name,
+            show_password = true,
+        }
+        local uri = config:instance_uri('peer', uri_opts)
+        if uri == nil then
+            local err = 'No suitable URI provided for instance %q'
+            error(err:format(instance_name), 0)
+        end
+
+        local conn_opts = {
+            wait_connected = false,
+            fetch_schema = opts.fetch_schema,
+        }
+        local ok, res = pcall(netbox.connect, uri, conn_opts)
+        if not ok then
+            local msg = 'Unable to connect to instance %q: %s'
+            error(msg:format(instance_name, res.message), 0)
+        end
+        conn = res
+        connections[instance_name] = conn
+    end
+
+    if opts.wait_connected == false then
+        return conn
+    end
+    -- wait_connected == true is not supported.
+    local wait_connected =
+        opts.wait_connected ~= true and opts.wait_connected or nil
+
+    assert(conn ~= nil)
+    local ok, res = pcall(conn.wait_connected, conn, wait_connected)
+    if not ok then
+        local msg = 'Unable to connect to instance %q: %s'
+        error(msg:format(instance_name, res.message), 0)
+    end
```

Comment:
First, I think that if something is not supported, it should lead to an error.

Next, it is a bit hard to understand that the `wait_connected` variable contains a timeout for creating a connection. I would check the type and assign it to a variable named as `connect_timeout`.

BTW, the API may be more straightforward if we add it as a separate `connect_timeout` option into the API. It is completely up to you, just my though.

Reply:
After discussion we decided to accept only boolean for `wait_connected` and add another numeric option `connect_timeout`. These options are passed directly to `net.box.connect()`. These two options works about the same way as `wait_connected` option for `cartridge.pool.connect()`. There is couple of cases when the behavior is different:
1) If `wait_connected` option for `cartridge.pool.connect()` more than default `connect_timeout` of `net.box.connect()`, than default `connect_timeout` is actually used;
2) If `wait_connected` option for `cartridge.pool.connect()` less than default `connect_timeout` of `net.box.connect()`, it was actually possible that an error returned before the connection is successfully established. This actually seems to be expected behavior, as the error says `Connection not established (yet)`.